radix-engine/src/model/epoch_manager/executables.rs

```diff
@@ -262,6 +287,126 @@ impl Executor for EpochManagerSetEpochExecutable {
     }
 }
 
+pub struct EpochManagerRegisterValidatorExecutable(RENodeId, EcdsaSecp256k1PublicKey);
+
+impl<W: WasmEngine> ExecutableInvocation<W> for EpochManagerRegisterValidatorInvocation {
+    type Exec = EpochManagerRegisterValidatorExecutable;
+
+    fn resolve<D: ResolverApi<W>>(
+        self,
+        deref: &mut D,
+    ) -> Result<(ResolvedActor, CallFrameUpdate, Self::Exec), RuntimeError>
+    where
+        Self: Sized,
+    {
+        let mut call_frame_update = CallFrameUpdate::empty();
+        let receiver = RENodeId::Global(GlobalAddress::System(self.receiver));
+        let resolved_receiver = deref_and_update(receiver, &mut call_frame_update, deref)?;
+
+        let actor = ResolvedActor::method(
+            NativeFn::EpochManager(EpochManagerFn::RegisterValidator),
+            resolved_receiver,
+        );
+        let executor =
+            EpochManagerRegisterValidatorExecutable(resolved_receiver.receiver, self.validator);
+
+        Ok((actor, call_frame_update, executor))
+    }
+}
+
+impl Executor for EpochManagerRegisterValidatorExecutable {
+    type Output = ();
+
+    fn execute<Y>(self, api: &mut Y) -> Result<((), CallFrameUpdate), RuntimeError>
+    where
+        Y: SystemApi + InvokableModel<RuntimeError>,
+    {
+        // TODO: Figure out how to move this access check into more appropriate place
+        {
+            let node_ids = api.get_visible_node_ids()?;
+            let auth_zone_id = node_ids
+                .into_iter()
+                .find(|n| matches!(n, RENodeId::AuthZoneStack(..)))
+                .expect("AuthZone does not exist");
+            let non_fungible_address = NonFungibleAddress::from_public_key(&self.1);
+            let access_rule = AccessRule::Protected(AccessRuleNode::ProofRule(ProofRule::Require(
+                SoftResourceOrNonFungible::StaticNonFungible(non_fungible_address),
+            )));
+
+            api.invoke(AuthZoneAssertAccessRuleInvocation {
+                receiver: auth_zone_id.into(),
+                access_rule,
+            })?;
+        }
+
+        let offset = SubstateOffset::EpochManager(EpochManagerOffset::PreparingValidatorSet);
+        let handle = api.lock_substate(self.0, offset, LockFlags::MUTABLE)?;
+        let mut substate_mut = api.get_ref_mut(handle)?;
+        substate_mut.validator_set().validator_set.insert(self.1);
```

Comment:
And we do not check existence?

Reply:
In this implementation it's implemented as idempotent call.